docs/advanced/integration_tests.md

```diff
@@ -0,0 +1,130 @@
+
+In order to make this GET call _Test_ToEuro_ does not need to concern itself about the IP on which the exchange application is running, nor its port. All _Test_ToEuro_ needs to know is the URL path of that endpoint, as the _client_ object does the rest.
+
+This _client_ object is provided out-of-the box by Gosoline, whenever a test is defined like this:
+
+[embedmd]:# (../../examples/getting_started/integration/api_test.go /func \(s \*ApiTestSuite\) SetupSuite/ /return definer.ApiDefiner\n}/)
+```go
+func (s *ApiTestSuite) SetupSuite() []suite.Option {
+	return []suite.Option{
+		suite.WithLogLevel("info"),
+		suite.WithConfigFile("../api/config.dist.yml"),
+		suite.WithFixtures(fixtureSets),
+		suite.WithClockProvider(s.clock),
+	}
+}
+
+func (s *ApiTestSuite) SetupApiDefinitions() apiserver.Definer {
+	return definer.ApiDefiner
+}
+```
+
+Implementing _SetupApiDefinitions_ when configuring your test will inform Gosoline that an API is being tested, thus the _client_ object will contain meaningful data. 
+
+### ApiServer integration tests
+
+The _client_ object is permissive, allowing access to the request itself. A simpler way to test an API, is provided by Gosoline's _ApiServerTestCase_. These type of tests also issue http calls and compare their responses against a predefined value, but they need to follow a predefined structure. Below we can see how the _resty client_ test above would look like as an _ApiServerTestCase_:
+
+[embedmd]:# (../../examples/getting_started/integration/api_test.go /func \(s \*ApiTestSuite\) Test_Euro/ /\n}/)
+```go
+func (s *ApiTestSuite) Test_Euro() *suite.ApiServerTestCase {
+	return &suite.ApiServerTestCase{
+		Method:             http.MethodGet,
+		Url:                "/euro/10/GBP",
+		Headers:            map[string]string{},
+		ExpectedStatusCode: http.StatusOK,
+		Assert: func(response *resty.Response) error {
+			s.Equal(http.StatusOK, response.StatusCode())
+
+			result, err := strconv.ParseFloat(string(response.Body()), 64)
+			s.NoError(err)
+			s.Equal(8.0, result)
+
+			return nil
+		},
+	}
+}
+```
+
+Notice the predefined structure of an _ApiServerTestCase_: it is an object that has fields for all the information needed in performing an http call, including a method that tests the result for correctness.
+
+### Stream integration tests
+
+If you have a Gosoline application that takes its input from a stream, you need a test which can run your application locally, send data to a stream, and check your application's output for correctness. 
+
+The `stream-consumer` application, found in `advanced/stream-consumer`, reads unsigned integers from an `inputEvent` sqs queue, increments them by one and publishes them to the `publisher-outputEvent` SNS topic. Both the SQS queue which provides it with input, and the SNS topic to which it publishes its results, are configured in its `config.dist.yml` file. Below is an extract from its integration test:
```

Comment:
`publisher-outputEvent` is an output, not an SNS topic. Only the configuration then connects that to an SQS queue and SNS topic, in the integration test those will be inMemory inputs/outputs